extensions/ql-vscode/src/view/results/AlertTablePathNodeRow.tsx

```diff
@@ -9,6 +17,151 @@ import { SarifLocation } from "./locations/SarifLocation";
+interface ModelTaxon {
+  location: SarifLogLocation;
+}
+
+/** Resolve an `ArtifactLocation` that might contain a relative reference instead of an absolute
+ * URI.
+ */
+function resolveArtifactLocation(
+  location: ArtifactLocation,
+  baseUri: URL,
+): ArtifactLocation {
+  if (location.uri === undefined) {
+    // No URI at all. Just return the original location.
+    return location;
+  }
+  return {
+    ...location,
+    uri: new URL(location.uri, baseUri).toString(),
+  };
+}
+
+/** Get the URI of the pack's local root directory, if available. */
+function getLocalPackUri(extension: ToolComponent): URL | undefined {
+  if (extension.locations === undefined) {
+    return undefined;
+  }
+
+  const localPackLocation = extension.locations.find(
+    (loc) =>
+      loc.properties !== undefined &&
+      loc.properties.tags !== undefined &&
+      loc.properties.tags.includes("CodeQL/LocalPackRoot"),
+  );
+  if (localPackLocation === undefined || localPackLocation.uri === undefined) {
+    return undefined;
+  }
+
+  return new URL(localPackLocation.uri);
+}
+
+/** Resolve a `ReportingDescriptorReference` to the `ReportingDescriptor` for the taxon that it
+ * refers to.
+ */
+function resolveTaxonDefinition(
+  run: Run,
+  taxonRef: ReportingDescriptorReference,
+): ModelTaxon | undefined {
+  const extensions = run.tool.extensions;
+  if (extensions === undefined) {
+    return undefined;
+  }
+
+  const extensionIndex = taxonRef.toolComponent?.index;
+  if (
+    extensionIndex === undefined ||
+    extensionIndex < 0 ||
+    extensionIndex >= extensions.length
+  ) {
+    return undefined;
+  }
+
+  const extension = extensions[extensionIndex];
+  if (extension.taxa === undefined) {
+    return undefined;
+  }
+
+  const localPackUri = getLocalPackUri(extension);
+  if (localPackUri === undefined) {
+    return undefined;
+  }
+
+  const taxonIndex = taxonRef.index;
+  if (
+    taxonIndex === undefined ||
+    taxonIndex < 0 ||
+    taxonIndex >= extension.taxa.length
+  ) {
+    return undefined;
+  }
+
+  const taxonDef = extension.taxa[taxonIndex];
+  if (taxonDef.properties === undefined) {
+    return undefined;
+  }
+
+  const location: PhysicalLocation =
+    taxonDef.properties["CodeQL/DataExtensionLocation"];
+  if (location === undefined || location.artifactLocation === undefined) {
+    return undefined;
+  }
+
+  return {
+    location: {
+      physicalLocation: {
+        ...location,
+        artifactLocation: resolveArtifactLocation(
+          location.artifactLocation,
+          localPackUri,
+        ),
+      },
+    },
+  };
+}
+
+/** Generate the React elements for each taxon. */
+function taxaLocations(
+  taxa: ReportingDescriptorReference[] | undefined,
+  run: Run | undefined,
+  onClick: () => void,
+) {
+  if (taxa === undefined || taxa.length === 0 || run === undefined) {
+    return [];
+  }
+
+  return taxa.flatMap((taxonRef, index) => {
```

Comment:
What are the situations where we will have multiple taxa for a given path element? Realistically, are there any degenerate cases where we can have lots of them and it can make the UI unreadable?

Either:

1. Ignore this problem since this is a canary-only feature
2. Only show the first X links and the user will need to find the rest directly from the the sarif
3. Show the first X links and add a "more" button that adds the remainder

Most likely, 1 is fine, but we will probably need to handle degenerate cases if this becomes a public feature.

Reply:
In practice, the number is almost always 1 or 2. The 2 case is usually due to having both a step model and a sink model on the same node. In theory I think there can be any number of models on a given node, most likely to happen if multiple MaD rows match the same AST element. @aschackmull might have a better idea of how likely this is to be a real problem.

In any case, I think we can leave this as-is unless it becomes a problem in practice, at least as long as it's still canary-only.

Reply:
I'm fine with that. I think you already mentioned to me that it would be at most two, but I'd forgotten.